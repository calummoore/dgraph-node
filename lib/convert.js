'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createTxnContext = createTxnContext;
exports.createLinRead = createLinRead;

var _protos = require('./protos');

var _protos2 = _interopRequireDefault(_protos);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createTxnContext({ startTs, keys, linRead }) {
  const txnContext = new _protos2.default.TxnContext();
  txnContext.keys = keys;
  txnContext.start_ts = startTs;
  txnContext.lin_read = createLinRead(linRead);
  return txnContext;
}

function createLinRead(linRead) {
  const lr = new _protos2.default.LinRead();
  lr.ids = linRead;
  return lr;
}