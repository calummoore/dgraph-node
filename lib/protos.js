'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; // @noflow


var _grpc = require('grpc');

var _grpc2 = _interopRequireDefault(_grpc);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const p1 = _grpc2.default.load(_path2.default.join(__dirname, '/../protos/api.proto'));
exports.default = _extends({}, p1.api);