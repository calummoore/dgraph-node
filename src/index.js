import DgraphClient from './dgraph'

export protos from './protos'
export default DgraphClient
