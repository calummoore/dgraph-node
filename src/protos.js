// @noflow
import grpc from 'grpc'
import path from 'path'

const p1 = grpc.load(path.join(__dirname, '/../protos/api.proto'))
export default { ...p1.api }
