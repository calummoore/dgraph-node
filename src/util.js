import reduce from 'lodash/reduce'

export function mergeMax (map, altMap) {
  return reduce(altMap, (mergedMap, value, key) => {
    // eslint-disable-next-line no-param-reassign
    mergedMap[key] = mergedMap[key]
      ? Math.max(mergedMap[key], parseFloat(value)) : parseFloat(value)
    return mergedMap
  }, map || {})
}
