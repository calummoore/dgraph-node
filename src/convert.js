import protos from './protos'

export function createTxnContext ({ startTs, keys, linRead }) {
  const txnContext = new protos.TxnContext()
  txnContext.keys = keys
  txnContext.start_ts = startTs
  txnContext.lin_read = createLinRead(linRead)
  return txnContext
}

export function createLinRead (linRead) {
  const lr = new protos.LinRead()
  lr.ids = linRead
  return lr
}
