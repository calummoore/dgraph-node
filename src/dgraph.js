import grpc from 'grpc'
import promisify from 'es6-promisify'
import protos from './protos'
import Transaction from './transaction'
import { mergeMax } from './util'
import { createLinRead } from './convert'

export default class DgraphClient {
  constructor (userConfig) {
    const config = {
      url: 'localhost:9080',
      debug: false,
      ...userConfig,
    }

    // Create gprc client into Dgraph
    const d = new protos.Dgraph(
      config.url,
      config.credentials || grpc.credentials.createInsecure(),
    )

    // Promisfy the proto fns
    this._mutate = promisify(d.mutate, d)
    this._query = promisify(d.query, d)
    this._alter = promisify(d.alter, d)
    this._commitOrAbort = promisify(d.commitOrAbort, d)

    this.linRead = null
    this.config = config
  }

  async query (query, linRead) {
    const queryRequest = new protos.Request()
    queryRequest.query = query
    if (linRead || this.linRead) queryRequest.lin_read = createLinRead(linRead || this.linRead)
    const resp = await this._query(queryRequest)
    this.updateContext(resp.txn)
    return {
      data: JSON.parse(resp.json.toString()),
      context: resp.txn,
    }
  }

  async mutate (mutation, commit = true, startTs) {
    const mutationRequest = new protos.Mutation()
    if (mutation.set) mutationRequest.set_nquads = Buffer.from(mutation.set, 'utf8')
    if (mutation.del) mutationRequest.del_nquads = Buffer.from(mutation.del, 'utf8')
    if (startTs) mutationRequest.start_ts = startTs
    mutationRequest.commit_now = commit
    const resp = await this._mutate(mutationRequest)
    return {
      data: {
        uids: resp.uids,
      },
      context: resp.context,
    }
  }

  async alter (schema) {
    const alterSchema = new protos.Operation()
    alterSchema.schema = schema
    return this._alter(alterSchema)
  }

  async dropAll () {
    const dropAllOp = new protos.Operation()
    dropAllOp.drop_all = true
    return this._alter(dropAllOp)
  }

  async dropAttr (attr) {
    const dropAttr = new protos.Operation()
    dropAttr.drop_atrr = attr
    return this._alter(dropAttr)
  }

  txn () {
    return new Transaction(this)
  }

  updateContext (context) {
    if (context.lin_read) {
      this.linRead = mergeMax(this.linRead, context.lin_read.ids)
    }
  }
}
