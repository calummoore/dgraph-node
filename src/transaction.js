import union from 'lodash/union'
// import protos from './protos'
import { mergeMax } from './util'
import { createTxnContext } from './convert'

export default class Transaction {
  constructor (client) {
    this.client = client
    this.startTs = null
    this.linRead = client.linRead
    this.keys = []
  }

  async query (query) {
    const resp = await this.client.query(query, this.linRead)
    this.updateContext(resp.context, this.linRead)
    return resp
  }

  async mutate (mutation) {
    const resp = await this.client.mutate(mutation, false, this.startTs)
    this.updateContext(resp.context)
    return resp
  }

  async commit () {
    const { startTs, linRead, keys } = this
    const txnContext = createTxnContext({
      startTs,
      linRead,
      keys,
    })
    return this.client._commitOrAbort(txnContext)
  }

  updateContext (context) {
    if (this.startTs === null) {
      this.startTs = context.start_ts
    }
    if (context.keys) {
      this.keys = union(context.keys.map(s => s.toString()), this.keys || [])
    }
    if (context.lin_read) {
      this.linRead = mergeMax(this.linRead, context.lin_read.ids)
    }
  }
}
