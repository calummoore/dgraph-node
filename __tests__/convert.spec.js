import { createLinRead, createTxnContext } from '../src/convert'
import protos from '../src/protos'

describe('convert', () => {
  it('should create a TxnContext object', () => {
    const txnContext = createTxnContext({
      keys: [1, 2, 3],
      startTs: 10,
      linRead: { 1: 10 },
    })
    expect(txnContext).toBeInstanceOf(protos.TxnContext)
    expect(txnContext.start_ts).toEqual(10)
    expect(txnContext.keys).toEqual([1, 2, 3])
    expect(txnContext.lin_read.ids).toEqual({ 1: 10 })
    expect(txnContext).toMatchSnapshot()
  })

  it('should create a LinRead object', () => {
    const linRead = createLinRead({
      1: 10,
    })
    expect(linRead).toBeInstanceOf(protos.LinRead)
    expect(linRead.ids).toEqual({ 1: 10 })
    expect(linRead).toMatchSnapshot()
  })
})
