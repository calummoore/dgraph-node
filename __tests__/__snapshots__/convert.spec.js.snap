// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert should create a LinRead object 1`] = `
 {
  "ids": Object {
    "1": 10,
  },
}
`;

exports[`convert should create a TxnContext object 1`] = `
 {
  "aborted": false,
  "commit_ts": Long {
    "high": 0,
    "low": 0,
    "unsigned": true,
  },
  "keys": Array [
    1,
    2,
    3,
  ],
  "lin_read":  {
    "ids": Object {
      "1": 10,
    },
  },
  "start_ts": 10,
}
`;
